crates/oci-distribution/src/client.rs

```diff
@@ -114,6 +127,73 @@ impl Client {
         })
     }
 
+    /// Push an image and return the uploaded URL of the image
+    ///
+    /// The client will check if it's already been authenticated and if
+    /// not will attempt to do.
+    ///
+    /// If a manifest is not provided, the client will attempt to generate
+    /// it from the provided image and config data.
+    ///
+    /// Returns pullable URL for the image
+    pub async fn push(
+        &mut self,
+        image_ref: &Reference,
+        image_data: &ImageData,
+        config_data: &[u8],
+        config_media_type: &str,
+        auth: &RegistryAuth,
+        manifest: Option<OciManifest>,
+    ) -> anyhow::Result<String> {
```

Comment:
I did find https://docs.rs/http/0.2.1/http/uri/index.html, which seems to be the standard type for that sort of thing. However there is also https://docs.rs/reqwest/0.10.9/reqwest/struct.Url.html which I think would be better if we wanted to use a type rather than just a string, since we use `reqwest` and each string URL is just converted into this struct anyways